packages/simple-cache-provider/src/SimpleCacheProvider.js

```diff
@@ -0,0 +1,181 @@
+      'Invalid resourceType: Expected a symbol, object, or function, but ' +
+        'instead received: %s. Strings and numbers are not permitted as ' +
+        'resource types.',
+      resourceType,
+    );
+
+    let recordCache: ?RecordCache<K, V>;
+    if (cache.has(resourceType)) {
+      recordCache = cache.get(resourceType);
+      if (recordCache.has(key)) {
+        return recordCache.get(key);
+      }
+    } else {
+      recordCache = new Map();
+      cache.set(resourceType, recordCache);
+    }
+
+    const record = {
+      status: Empty,
+      suspender: null,
+      value: null,
+      error: null,
+    };
+    recordCache.set(key, record);
+    return record;
+  }
+
+  function load(record, key, miss) {
+    const suspender = miss(key);
+    record.suspender = suspender;
+    suspender.then(
+      value => {
+        // Resource loaded successfully.
+        record.suspender = null;
+        record.status = Resolved;
+        record.value = value;
+      },
+      error => {
+        // Resource failed to load. Stash the error for later so we can throw it
+        // the next time it's requested.
+        record.suspender = null;
+        record.status = Rejected;
+        record.error = error;
+      },
+    );
+    return suspender;
+  }
+
+  return {
+    $$typeof: CACHE_TYPE,
+    invalidate() {
+      invalidator();
+    },
+    preload<K, V>(resourceType: any, key: K, miss: K => Promise<V>): void {
+      const record: Record<K, V> = getRecord(resourceType, key);
+      const suspender = record.suspender;
+      if (suspender !== null) {
+        // There's already a pending request.
+        return;
+      }
+      switch (record.status) {
+        case Empty:
+          // Warm the cache.
+          load(record, key, miss);
+          return;
+        case Resolved:
+        case Rejected:
+          // The resource is already in the cache.
+          return;
+      }
+    },
+    read<K, V>(resourceType: any, key: K, miss: K => Promise<V>): V {
+      const record: Record<K, V> = getRecord(resourceType, key);
+      const suspender = record.suspender;
+      if (suspender !== null) {
+        // There's already a pending request.
+        throw suspender;
+      }
+      switch (record.status) {
+        case Empty:
+          // Load the requested resource.
+          throw load(record, key, miss);
+        case Resolved:
+          return record.value;
+        case Rejected:
+          // The requested resource previously failed loading.
+          // TODO: When should we try again?
+          throw record.error;
+      }
+    },
+  };
+}
+
+type ResourceReader<K, V> = K => V;
+
+type Resource<K, V> = ResourceReader<K, V, H> & {|
+  preload(key: K): void,
+|};
+
+export function createResource<K, V, H>(
+  loadResource: K => V,
+  hash?: K => H = identity,
+): Resource<K, V> {
+  // The read function itself serves as the resource type.
+  function read(cache, key) {
+    invariant(
```

Comment:
Slow. This is a suuuuper hot path. Make it `__DEV__` only (and therefore the whole $$typeof can be dev only).